packages/lodestar/src/chain/blocks/process.ts

```diff
@@ -125,28 +130,52 @@ export function updateForkChoice(
   return forkChoice.headBlockRoot();
 }
 
+export function emitCheckpointEvent(emitter: ChainEventEmitter, checkpointStateContext: ITreeStateContext): void {
+  const config = checkpointStateContext.epochCtx.config;
+  const slot = checkpointStateContext.state.slot;
+  assert.true(slot % config.params.SLOTS_PER_EPOCH === 0, "Checkpoint state slot must be first in an epoch");
+  emitter.emit(
+    "checkpoint",
+    {
+      root: config.types.BeaconBlockHeader.hashTreeRoot(checkpointStateContext.state.latestBlockHeader),
+      epoch: computeEpochAtSlot(config, slot),
+    },
+    checkpointStateContext
+  );
+}
+
 export async function runStateTransition(
-  config: IBeaconConfig,
-  db: IBeaconDb,
-  logger: ILogger,
-  stateContext: Required<IStateContext>,
+  emitter: ChainEventEmitter,
+  stateContext: ITreeStateContext,
   job: IBlockProcessJob
 ): Promise<IStateContext | null> {
+  const config = stateContext.epochCtx.config;
+  const {SLOTS_PER_EPOCH} = config.params;
   try {
+    const preSlot = stateContext.state.slot;
+    const postSlot = job.signedBlock.message.slot;
+    const preEpoch = computeEpochAtSlot(config, preSlot);
+    for (
+      let nextEpochSlot = computeStartSlotAtEpoch(config, preEpoch + 1);
+      nextEpochSlot < postSlot;
+      nextEpochSlot += SLOTS_PER_EPOCH
+    ) {
+      processSlots(stateContext.epochCtx, stateContext.state, nextEpochSlot);
+      emitCheckpointEvent(emitter, stateContext);
+    }
     // if block is trusted don't verify proposer or op signature
-    return fastStateTransition(stateContext, job.signedBlock, {
+    const postStateContext = fastStateTransition(stateContext, job.signedBlock, {
       verifyStateRoot: true,
       verifyProposer: !job.trusted,
       verifySignatures: !job.trusted,
-    });
+    }) as ITreeStateContext;
+    if (postSlot % SLOTS_PER_EPOCH === 0) {
+      emitCheckpointEvent(emitter, postStateContext);
```

Comment:
I suppose `postStateContext.state.latestBlockHeader.stateRoot` is ZERO_HASH so we can't use `emitCheckpointEvent` here.

The old emitter code of getting root from `config.types.BeaconBlock.hashTreeRoot(job.signedBlock.message)` should work fine.

Reply:
we never run state transition against the genesis block so we should be good here :)

Reply:
I refer to this https://github.com/ChainSafe/lodestar/blob/master/packages/lodestar-beacon-state-transition/src/block/blockHeader.ts#L24

my understanding is `postStateContext.state.latestBlockHeader.stateRoot` is always ZERO_HASH after we run a state transition since we don't know stateRoot at that time. Next `stateTransition` or `processSlot` should set previous `blockRoot` to state